storm-core/src/jvm/org/apache/storm/pacemaker/Pacemaker.java

```diff
@@ -0,0 +1,246 @@
+        if (authenticated) {
+            Set<String> pulseIds = new HashSet<>();
+            for (String key : heartbeats.keySet()) {
+                String[] replaceStr = key.replaceFirst(path, "").split("/");
+                String trimmed = null;
+                for (String str : replaceStr) {
+                    if (!str.equals("")) {
+                        trimmed = str;
+                        break;
+                    }
+                }
+                if (trimmed != null && key.indexOf(path) == 0) {
+                    pulseIds.add(trimmed);
+                }
+            }
+            HBMessageData hbMessageData = HBMessageData.nodes(new HBNodes(new ArrayList(pulseIds)));
+            return new HBMessage(HBServerMessageType.GET_ALL_NODES_FOR_PATH_RESPONSE, hbMessageData);
+        } else {
+            return notAuthorized();
+        }
+    }
+
+    private HBMessage getPulse(String path, boolean authenticated) {
+        if (authenticated) {
+            byte[] details = heartbeats.get(path);
+            LOG.debug("Getting Pulse for path [ {} ]...data [ {} ].", path, details);
+            pacemakerStats.getPulseCount.incrementAndGet();
+            if (details != null) {
+                pacemakerStats.totalSentSize.addAndGet(details.length);
+            }
+            HBPulse hbPulse = new HBPulse();
+            hbPulse.set_id(path);
+            hbPulse.set_details(details);
+            return new HBMessage(HBServerMessageType.GET_PULSE_RESPONSE, HBMessageData.pulse(hbPulse));
+        } else {
+            return notAuthorized();
+        }
+    }
+
+    private HBMessage deletePath(String path) {
+        String prefix = path.endsWith("/") ? path : (path + "/");
+        for (String key : heartbeats.keySet()) {
+            if (key.indexOf(prefix) == 0)
+                deletePulseId(key);
+        }
+        return new HBMessage(HBServerMessageType.DELETE_PATH_RESPONSE, null);
+    }
+
+    private HBMessage deletePulseId(String path) {
+        LOG.debug("Deleting Pulse for id [ {} ].", path);
+        heartbeats.remove(path);
+        return new HBMessage(HBServerMessageType.DELETE_PULSE_ID_RESPONSE, null);
+    }
+
+    private void updateLargestHbSize(int size) {
+        int newValue = size;
+        while (true) {
+            int oldValue = pacemakerStats.largestHeartbeatSize.get();
+            if (newValue > oldValue) {
+                if (!pacemakerStats.largestHeartbeatSize.compareAndSet(oldValue, newValue))
+                    continue;
+            }
+            break;
+        }
+    }
+
+    private void updateAverageHbSize(int size) {
+        while (true) {
+            int oldValue = pacemakerStats.averageHeartbeatSize.get();
+            int count = pacemakerStats.sendPulseCount.get();
+            int newValue = ((count * oldValue) + size) / (count + 1);
+            if (!pacemakerStats.averageHeartbeatSize.compareAndSet(oldValue, newValue))
```

Comment:
Minor. It can be simplified to 
```
if (pacemakerStats.averageHeartbeatSize.compareAndSet(oldValue, newValue))
 break;
```
Basically you can check for true condition and then break.